File: batch-crawl-photos.js
const { chromium } = require('playwright');
const fs = require('fs');

async function crawlVenuePhotos(url, venueName) {
  const browser = await chromium.launch({
    headless: true
  });

  const context = await browser.newContext({
    userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
  });

  const page = await context.newPage();

  try {
    await page.goto(url, {
      waitUntil: 'domcontentloaded',
      timeout: 20000
    });

    await page.waitForTimeout(3000);

    const photos = await page.evaluate(() => {
      const photos = [];
      const images = document.querySelectorAll('img');

      images.forEach(img => {
        if (img.src && !img.src.includes('logo') && !img.src.includes('icon') && !img.src.includes('loading')) {
          if (img.width > 100 && img.height > 100) {
            photos.push(img.src);
          }
        }
      });

      return photos;
    });

    await browser.close();

    return {
      venueName,
      url,
      photos: photos.slice(0, 10),
      success: true
    };

  } catch (error) {
    await browser.close();
    return {
      venueName,
      url,
      error: error.message,
      photos: [],
      success: false
    };
  }
}

async function main() {
  console.log('🚀 批次爬蟲更新場地照片！\n');

  const venues = JSON.parse(fs.readFileSync('sample-data.json', 'utf8'));

  // 找出需要更新照片的場地
  const needsUpdate = venues.filter(v =>
    !v.images?.main ||
    v.images?.main?.includes('unsplash') ||
    v.images?.main?.includes('wikipedia') ||
    v.images?.main?.includes('upload.wikimedia')
  );

  console.log(`📊 需要更新的場地: ${needsUpdate.length} 個\n`);

  // 只處理前 50 個
  const toProcess = needsUpdate.slice(0, 50);
  console.log(`🎯 本次處理: ${toProcess.length} 個\n`);

  const results = [];

  for (const venue of toProcess) {
    const fullName = `${venue.name} ${venue.roomName || ''}`.trim();
    console.log(`🔍 爬蟲: ${fullName}`);
    console.log(`   URL: ${venue.url}`);

    const result = await crawlVenuePhotos(venue.url, fullName);
    results.push(result);

    if (result.success && result.photos.length > 0) {
      console.log(`   ✅ 成功: ${result.photos.length} 張照片\n`);
    } else {
      console.log(`   ❌ 失敗: ${result.error || '無照片'}\n`);
    }

    // 每完成 10 個保存一次
    if (results.length % 10 === 0) {
      fs.writeFileSync('batch-crawl-progress.json', JSON.stringify(results, null, 2));
    }
  }

  // 保存最終結果
  fs.writeFileSync('batch-crawl-results.json', JSON.stringify(results, null, 2));
  console.log('✅ 已保存到 batch-crawl-results.json\n');

  // 統計
  const success = results.filter(r => r.success && r.photos.length > 0).length;
  const failed = results.filter(r => !r.success || r.photos.length === 0).length;

  console.log('📊 統計:\n');
  console.log(`成功: ${success}/${toProcess.length}`);
  console.log(`失敗: ${failed}/${toProcess.length}\n`);

  console.log('🎉 批次爬蟲完成！');
}

main().catch(console.error);

File: batch_crawl_photos.py
#!/usr/bin/env python3
"""
批量爬取場地官網照片
"""

import json
import re
import requests
from urllib.parse import urljoin, urlparse
from bs4 import BeautifulSoup
import time

# 載入場地資料
with open('./venues-all-cities.json', 'r', encoding='utf-8') as f:
    all_venues = json.load(f)

# 載入需要更新的清單
with open('./batch-update-list.json', 'r', encoding='utf-8') as f:
    update_list = json.load(f)

# 建立場地 ID 對照表
venue_map = {v['id']: v for v in all_venues}

def crawl_photos(url, venue_name):
    """從官網爬取照片"""
    print(f"  正在爬取: {venue_name}")
    print(f"  URL: {url}")
    
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
    }
    
    try:
        response = requests.get(url, headers=headers, timeout=10)
        response.raise_for_status()
        
        soup = BeautifulSoup(response.text, 'html.parser')
        
        # 尋找所有圖片
        images = []
        
        # 1. 找 <img> 標籤
        for img in soup.find_all('img'):
            src = img.get('src') or img.get('data-src') or img.get('data-lazy-src')
            if src:
                full_url = urljoin(url, src)
                # 過濾掉明顯不是場地照片的圖片
                if not any(x in full_url.lower() for x in ['logo', 'icon', 'favicon', 'avatar', 'button', 'arrow']):
                    images.append(full_url)
        
        # 2. 找背景圖片 (style="background-image: url(...)")
        for element in soup.find_all(style=re.compile(r'background-image')):
            style = element.get('style', '')
            match = re.search(r'url\(["\']?([^"\')\s]+)["\']?\)', style)
            if match:
                bg_url = urljoin(url, match.group(1))
                if not any(x in bg_url.lower() for x in ['logo', 'icon', 'favicon']):
                    images.append(bg_url)
        
        # 去重
        images = list(dict.fromkeys(images))
        
        print(f"  找到 {len(images)} 張圖片")
        return images[:10]  # 最多取 10 張
        
    except Exception as e:
        print(f"  ❌ 錯誤: {e}")
        return []

# 執行批量爬取
results = []
for item in update_list:
    venue_id = item['id']
    venue = venue_map.get(venue_id)
    
    if not venue:
        print(f"找不到場地 ID: {venue_id}")
        continue
    
    print(f"\n[{venue_id}] {venue['name']}")
    
    # 嘗試爬取官網
    url = venue.get('url') or venue.get('meetingPageUrl')
    if not url:
        print(f"  ⚠️ 沒有官網 URL")
        continue
    
    photos = crawl_photos(url, venue['name'])
    
    if photos:
        results.append({
            'id': venue_id,
            'name': venue['name'],
            'url': url,
            'photos': photos,
            'main': photos[0] if photos else None,
            'gallery': photos[1:] if len(photos) > 1 else []
        })
    
    time.sleep(1)  # 避免過度請求

# 輸出結果
print(f"\n\n=== 爬取結果 ===")
print(f"成功: {len(results)} / {len(update_list)}")

with open('./crawled-photos-result.json', 'w', encoding='utf-8') as f:
    json.dump(results, f, ensure_ascii=False, indent=2)

print(f"\n已輸出到 crawled-photos-result.json")

# 顯示結果摘要
for r in results:
    print(f"\n[{r['id']}] {r['name']}")
    print(f"  主照片: {r['main']}")
    print(f"  相簿: {len(r['gallery'])} 張")
